pkg/healthchecker/health_checker.go

```diff
@@ -0,0 +1,146 @@
+	// The bash commands to kill the process will fail if the service is down and hence ignore.
+	repairFunc         func()
+	crictlPath         string
+	healthCheckTimeout time.Duration
+	coolDownTime       time.Duration
+}
+
+// NewHealthChecker returns a new health checker configured with the given options.
+func NewHealthChecker(hco *options.HealthCheckerOptions) (types.HealthChecker, error) {
+	hc := &healthChecker{
+		enableRepair:       hco.EnableRepair,
+		crictlPath:         hco.CriCtlPath,
+		healthCheckTimeout: hco.HealthCheckTimeout,
+		coolDownTime:       hco.CoolDownTime,
+	}
+	switch hco.Component {
+	case types.KubeletComponent:
+		hc.healthCheckFunc = kubeletHealthCheck
+		hc.repairFunc = kubeletRepair
+	case types.ContainerRuntimeComponent:
+		hc.healthCheckFunc = getContainerRuntimeHealthCheck(hco.ContainerRuntime, hco.CriCtlPath)
+		hc.repairFunc = getContainerRuntimeRepair(hco.ContainerRuntime)
+	default:
+		return nil, fmt.Errorf("invalid component : %+v specified in health check options", hco.Component)
+	}
+	return hc, nil
+}
+
+// CheckHealth checks for the health of the component and tries to repair if enabled.
+// Returns true if healthy, false otherwise.
+func (hc *healthChecker) CheckHealth() bool {
+	// Poll till the health check timeout for the component to be up.
+	if err := wait.PollImmediate(hc.healthCheckTimeout, hc.healthCheckTimeout, func() (bool, error) {
+		healthy := hc.healthCheckFunc()
+		if healthy {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		// The service is unhealthy.
+		// Attempt repair based on flag.
+		if hc.enableRepair {
+			glog.Infof("health-checker: component is unhealthy, proceeding to repair")
+			hc.repairFunc()
+			// stall for cool down period after repairing
+			time.Sleep(hc.coolDownTime)
+		}
+		return false
+	}
+	return true
+}
+
+// kubeletHealthCheck checks the health of kubelet by hitting to the health endpoint.
+// Returns true if healthy, false otherwise.
+func kubeletHealthCheck() bool {
+	response, err := http.Get(types.KubeletHealthCheckEndpoint)
+	if err != nil || response.StatusCode != http.StatusOK {
+		return false
+	}
+	return true
+}
+
+// kubeletRepair tries to kill the kubelet service if a problem is detected.
+func kubeletRepair() {
+	execCommand("systemctl", "kill", "kubelet")
+}
+
+// getContainerRuntimeHealthCheck returns the health check function depending on the runtime.
+func getContainerRuntimeHealthCheck(runtime, crictlPath string) func() bool {
+	if runtime == types.DockerRuntime {
+		// Use "docker ps" for docker health check. Not using crictl for docker to remove
+		// dependency on the kubelet.
+		return func() bool {
+			if err := execCommand("docker", "ps"); err != nil {
+				return false
+			}
+			return true
+		}
+	}
+	// Use "crictl pods" for containerd health check.
+	return func() bool {
+		if err := execCommand(crictlPath, "pods"); err != nil {
```

Comment:
Where is the CRI socket path configured? Are we going to assume it is configured on the node?

Reply:
Added a flag to specify this.